snuba/web/rpc/v1/resolvers/R_eap_spans/common/sampling_in_storage_util.py

```diff
@@ -1,23 +1,127 @@
+        )
+        if estimated_query_duration_to_this_tier <= SENTRY_TIMEOUT - ERROR_BUDGET:
+            target_tier = tier
+    return target_tier
+
+
+def is_best_effort_mode(in_msg: T) -> bool:
+    return (
+        in_msg.meta.HasField("downsampled_storage_config")
+        and in_msg.meta.downsampled_storage_config.mode
+        == DownsampledStorageConfig.MODE_BEST_EFFORT
+    )
 
 
-def construct_query_settings(
-    in_msg: TimeSeriesRequest | TraceItemTableRequest,
-) -> HTTPQuerySettings:
-    query_settings = (
-        setup_trace_query_settings() if in_msg.meta.debug else HTTPQuerySettings()
+def build_snuba_request(
+    request: T, query_settings: QuerySettings, build_query: Callable[[T], Query]
+) -> SnubaRequest:
+    return SnubaRequest(
+        id=uuid.UUID(request.meta.request_id),
+        original_body=MessageToDict(request),
+        query=build_query(request),
+        query_settings=query_settings,
+        attribution_info=AttributionInfo(
+            referrer=request.meta.referrer,
+            team="eap",
+            feature="eap",
+            tenant_ids={
+                "organization_id": request.meta.organization_id,
+                "referrer": request.meta.referrer,
+            },
+            app_id=AppID("eap"),
+            parent_api="eap_span_samples",
+        ),
     )
-    if in_msg.meta.HasField("downsampled_storage_config"):
-        if (
-            in_msg.meta.downsampled_storage_config.mode
-            == DownsampledStorageConfig.MODE_PREFLIGHT
-        ):
-            query_settings.set_sampling_tier(Tier.TIER_512)
-
-    return query_settings
+
+
+def run_query_to_correct_tier(
+    in_msg: T,
+    query_settings: HTTPQuerySettings,
+    timer: Timer,
+    build_query: Callable[[T], Query],
+) -> QueryResult:
+    if not in_msg.meta.HasField("downsampled_storage_config"):
+        return run_query(
+            dataset=PluggableDataset(name="eap", all_entities=[]),
+            request=build_snuba_request(in_msg, query_settings, build_query),
+            timer=timer,
+        )
+
+    query_settings.set_sampling_tier(Tier.TIER_512)
+
+    request_to_most_downsampled_tier = build_snuba_request(
+        in_msg, query_settings, build_query
+    )
+
+    if is_best_effort_mode(in_msg):
+        query_settings.set_record_query_duration(True)
+
+    res = run_query(
+        dataset=PluggableDataset(name="eap", all_entities=[]),
+        request=request_to_most_downsampled_tier,
+        timer=timer,
+    )
+
+    if is_best_effort_mode(in_msg):
+        query_settings.push_clickhouse_setting(
+            "max_execution_time", SENTRY_TIMEOUT / 1000 - ERROR_BUDGET
+        )
+        query_settings.push_clickhouse_setting("timeout_overflow_mode", "break")
+        query_settings.set_sampling_tier(_get_target_tier(timer))
+        query_settings.set_record_query_duration(
+            False
+        )  # doesn't make a functional diff
```

Comment:
what about the case where so much data is scanned in the lowest tier that we have no budget left?